### PR TITLE
[MODINVOSTO-162] Create invoice line numbers based on new invoice field

### PR DESCRIPTION
**Pull request by damien-git**
## Purpose
[MODINVOSTO-162](https://issues.folio.org/browse/MODINVOSTO-162) - Jumps in invoice line numbers

## Approach
- Removed existing sequence-based implementation
- Replaced by using the new `nextInvoiceLineNumber` field in invoices
- Created a new service for invoice line numbers
- Added a way to create `nextInvoiceLineNumber` based on the invoice lines if it is not defined
- Created new `InvoiceLineDAO` to read invoice lines
- Added a migration script
- Updated unit tests

This needs to be reviewed and merged together with the related PRs in [mod-invoice](https://github.com/folio-org/mod-invoice/pull/412) and [mod-orders](https://github.com/folio-org/mod-orders/pull/750).

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all the appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?
- Did you modify code to call some additional endpoints?
  - [ ] If so, do you check that necessary module permission added in ModuleDescriptor-template.yaml file?

Ideally, all the PRs involved in breaking changes would be merged on the same day
to avoid breaking the folio-testing environment.
Communication is paramount if that is to be achieved,
especially as the number of inter-module and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems,
ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/dao/invoice/InvoicePostgresDAO.java [hunk 1]
```diff
@@ -101,6 +70,14 @@ public Future<DBClient> createInvoice(Invoice invoice, DBClient client) {
     return promise.future();
   }
 
+  @Override
+  public Future<Void> updateInvoice(Invoice invoice, Conn conn) {
+    return conn.update(INVOICE_TABLE, invoice, invoice.getId())
+      .onFailure(t -> log.error("updateInvoice failed for invoice with id {}", invoice.getId(), t))
```
Comment (grabsefx) [inline src/main/java/org/folio/dao/invoice/InvoicePostgresDAO.java:76]: `updateInvoice` will always return succeded future in this case. Possible exception occured while updating invoice will be just logged and ignored. Is this expected behavior?

Comment (damien-git) [reply src/main/java/org/folio/dao/invoice/InvoicePostgresDAO.java:76]: I don't understand. `onFailure` does not prevent a failed future from being returned (like `recover`), it is just an event handler for when it fails...

**Review by SerhiiNosko — APPROVED** [review]

File: src/main/java/org/folio/dao/invoice/InvoicePostgresDAO.java (hunk 1 above)
Comment (grabsefx) [reply src/main/java/org/folio/dao/invoice/InvoicePostgresDAO.java:76]: right. sorry for confusing

**Review by grabsefx — APPROVED** [review]

File: src/main/java/org/folio/service/InvoiceLineNumberService.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  private int getLastLineNumber(List<InvoiceLine> invoiceLines) {
+    try {
+      return invoiceLines.stream()
+        .map(InvoiceLine::getInvoiceLineNumber)
+        .filter(Objects::nonNull)
+        .map(Integer::parseInt)
+        .reduce(Integer::max)
+        .orElse(0);
+    } catch (Exception t) {
+      return 0;
```
Comment (Abdulkhakimov) [inline src/main/java/org/folio/service/InvoiceLineNumberService.java:100]: can we add loggin here?

**Review by Abdulkhakimov — APPROVED** [review]

File: src/main/java/org/folio/service/InvoiceLineNumberService.java (hunk 2 above)
Comment (damien-git) [reply src/main/java/org/folio/service/InvoiceLineNumberService.java:101]: Yes.

**Pull request merged**